### ISSUE TITLE
Error output differs depending on whether the source is a file or directory.

### ISSUE THREAD

**Issue by robwierzbowski** [sass/sass#1715]
Errors are reported on different channels and with different text depending on whether Sass is compiling a file or updating a directory. This makes it really hard to write tools that integrate with Sass (like gulp-ruby-sass).

On a directory:

```
// command: sass source:dest
stdout:       error source/fixture-a.scss (Line 4: Invalid CSS after "$base-font-size ": expected ":", was "16px !default;")

stdout:       error source/fixture-error.scss (Line 1: File to import not found or unreadable: i-dont-exist.
Load path: /Users/robw/Documents/Contrib/xxx)
```

On a file:

```
// command: sass source/fixture-a.scss dest/fixture-a.css
stderr: Error: Invalid CSS after "$base-font-size ": expected ":", was "16px !default;"
        on line 4 of source/fixture-a.scss
  Use --trace for backtrace.

// command: sass source/fixture-error.scss dest/fixture-error.css
stderr: Error: File to import not found or unreadable: i-dont-exist.
       Load path: /Users/robw/Documents/Contrib/xxx
        on line 1 of source/fixture-error.scss
  Use --trace for backtrace.

```

Is it possible for these errors to use consistent copy on consistent output channels?

**Comment by robwierzbowski** [comment]
Errors are sometimes mixed in with other output, sometimes part of their own stdout/stderr emission. I'm not sure if this is an artifact of the (node) app I'm using to listen to streams or if it is part of Sass logging. Here's the output separated with timestamps of the same task run twice in a row:

```
// Run one
[15:24:50] 
directory ./

[15:24:50] 
source/fixture-a.scss:11 DEBUG: This is a debug statement
WARNING: This is a warn statement
         on line 12 of source/fixture-a.scss

[15:24:50] 
write ./fixture-a.css
      write ./fixture-a.css.map

[15:24:50] error source/fixture-error.scss (Line 1: File to import not found or unreadable: i-dont-exist.
Load path: /Users/robw/Documents/Contrib/gulp-ruby-sass/fixture/source)

[15:24:50] 
directory ./nested

[15:24:50] 
write ./nested/fixture-b.css
      write ./nested/fixture-b.css.map
```

```
// Run two
[15:24:54] 
directory ./

[15:24:54] 
source/fixture-a.scss:11 DEBUG: This is a debug statement
WARNING: This is a warn statement
         on line 12 of source/fixture-a.scss

[15:24:54] 
write ./fixture-a.css
      write ./fixture-a.css.map
      error source/fixture-error.scss (Line 1: File to import not found or unreadable: i-dont-exist.
Load path: /Users/robw/Documents/Contrib/gulp-ruby-sass/fixture/source)

[15:24:54] 
directory ./nested

[15:24:54] 
write ./nested/fixture-b.css

[15:24:54] 
write ./nested/fixture-b.css.map
```

The error is part of another message about 50% of the times I run the Sass command. This type of mixing error logs with non-error logs makes it very hard to write tools that interact with Sass errors.

**Comment by nex3** [comment]
In general, I'd encourage tool authors to use the Ruby APIs for fine-grained control of and information about Sass. That's always going to give you more power, flexibility, and comprehensibility than trying to parse the command-line tool's output; even if the only thing you do is export the APIs via JSON or whatever, you'll have full control of what gets output and how. Ultimately the command-line tool is geared towards human users, so its output won't be entirely standardized when that's not what's most helpful for humans.

For your specific issues:
- Errors are emitted on stdout when doing `--watch` because they are in a sense part of its "expected output". It emits messages for various status updates, and errors are just one sort of those, so they don't use a separate stream. `--update` does the same because it's basically the "single-run" version of `--watch`.
- Similarly, since their job is to emit a large number of status updates, `--watch` and `--update` use a terser format for their errors than a single-file compilation, which wants errors to be eye-catching.
- I have no idea what's going on with your "merged messages". We emit those separately internally, although I'm not sure whether POSIX guarantees that separate writes to the print buffer will always be visible as separate packets to an external process.

**Comment by robwierzbowski** [comment]
Thanks for the reply, I appreciate the detail. Where would I find the docs
for the Ruby API?

On Friday, May 8, 2015, Natalie Weizenbaum notifications@github.com wrote:

> Closed #1715 https://github.com/sass/sass/issues/1715.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/sass/sass/issues/1715#event-300728212.

## 

Rob Wierzbowski
@robwierzbowski http://twitter.com/#!/robwierzbowski
http://github.com/robwierzbowski
http://robwierzbowski.com

**Comment by nex3** [comment]
[It's all here](http://sass-lang.com/documentation/).
